dev_tools/oneapi/add_host_function.sh

```diff
@@ -0,0 +1,305 @@
+                    echo "static_assert(false, \"Only inferred type are allowed after Config\");"
+                fi
+            else
+                if [ "${rev_id_list[${identification}]}" -ge "${max_num}" ]; then
+                    max_num=${rev_id_list[${identification}]}
+                    if [ -n "${list_list[${identification}]}" ]; then
+                        parameter_name=", ${parameter_name}"
+                    fi
+                    list_list[${identification}]="${list_list[${identification}]}${parameter_name}"
+                else
+                    echo "static_assert(false, \"Need to follow bool, int, size_type, type ordering\");"
+                fi
+            fi
+        fi
+        temp=""
+    done
+    # for item in "${!list_list[@]}"; do
+    #     echo "//$item - ${list_list[$item]}"
+    # done
+    config_generation=""
+    if [[ "${default_config}" = *"config_set"* ]]; then
+        # When the config is set in template, use it directly.
+        config_generation="::gko::syn::value_list<Config, ${default_config}>();"
+    else
+        # When config use the constexpr variable, use the variable name. The constexpr variable will be changed to Config_list afterwards.
+        config_generation="${default_config}_list"
+    fi
+    echo "template <${template_rm_config}>
+        void ${name}${CONFIG_SELECTION_SUFFIX}
+        (dim3 grid, dim3 block, size_t dynamic_shared_memory, cudaStream_t stream, std::shared_ptr<const CudaExecutor> exec, ${variable}) {
+        auto config_list = ${config_generation};
+        auto exec_config = exec->get_config();
+        ${name}${CONFIG_SELECTION_SUFFIX}(
+            config_list,
+            [&exec_config] (Config config) {
+                return (get_warp_size(exec_config) == get_warp_size(config)) &&
+                        (get_block_size(exec_config) >= get_block_size(config));
+                },
+            ::gko::syn::value_list<${list_list[bool]}>(),
+            ::gko::syn::value_list<${list_list[int]}>(),
+            ::gko::syn::value_list<${list_list[size_type]}>(),
+            ::gko::syn::type_list<${list_list[typename]}>(),
+            grid, block, dynamic_shared_memory, stream, ${var_input});
+    }
+    "
+}
+
+
+GLOBAL_KEYWORD="__global__"
+TEMPLATE_REGEX="^ *template <*"
+FUNCTION_START="^ *(template *<|${GLOBAL_KEYWORD}|void)"
+FUNCTION_NAME_END=".*\{.*"
+SCOPE_START="${FUNCTION_NAME_END}"
+SCOPE_END=".*\}.*"
+CHECK_GLOBAL_KEYWORD=".*${GLOBAL_KEYWORD}.*"
+FUNCTION_HANDLE=""
+DURING_FUNCNAME="false"
+ANAYSIS_FUNC=" *(template *<(.*)>)?.* (.*)\((.*)\)"
+START_BLOCK_REX="^( *\/\*| *\/\/)"
+END_BLOCK_REX="\*\/$| *\/\/"
+IN_BLOCK=0
+IN_FUNC=0
+STORE_LINE=""
+STORE_REGEX="__ *$"
+HOST_SUFFIX="_AUTOHOSTFUNC"
+EXTRACT_KERNEL="false"
+GINKGO_LICENSE_BEACON="******************************<GINKGO LICENSE>******************************"
+DURING_LICENSE="false"
+SKIP="false"
+MAP_FILE="map_list"
+rm "${MAP_FILE}"
```

Comment:
Some of the previous definitions can be shared as well, in particular the LICENSE stuff